sdks/java/harness/src/main/java/org/apache/beam/fn/harness/AssignWindowsRunner.java

```diff
@@ -0,0 +1,157 @@
+        Map<String, Coder> coders,
+        Map<String, WindowingStrategy> windowingStrategies,
+        Multimap<String, FnDataReceiver<WindowedValue<?>>> pCollectionIdsToConsumers,
+        Consumer<ThrowingRunnable> addStartFunction,
+        Consumer<ThrowingRunnable> addFinishFunction)
+        throws IOException {
+      checkArgument(
+          PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN.equals(pTransform.getSpec().getUrn()));
+      checkArgument(pTransform.getInputsCount() == 1, "Expected only one input");
+      checkArgument(pTransform.getOutputsCount() == 1, "Expected only one output");
+      WindowIntoPayload payload = WindowIntoPayload.parseFrom(pTransform.getSpec().getPayload());
+
+      WindowFn<?, ?> windowFn =
+          WindowingStrategyTranslation.windowFnFromProto(payload.getWindowFn());
+
+      String outputId = Iterables.getOnlyElement(pTransform.getOutputsMap().values());
+      Collection<FnDataReceiver<WindowedValue<?>>> outputConsumers =
+          pCollectionIdsToConsumers.get(outputId);
+      AssignWindowsRunner<?, ?> runner = AssignWindowsRunner.create(windowFn, outputConsumers);
+
+      String inputId = Iterables.getOnlyElement(pTransform.getInputsMap().values());
+      pCollectionIdsToConsumers.put(inputId, runner.asReceiver());
+      return runner;
+    }
+  }
+
+  /** A registrar which provides a factory to handle Java {@link WindowFn WindowFns}. */
+  @AutoService(PTransformRunnerFactory.Registrar.class)
+  public static class Registrar implements PTransformRunnerFactory.Registrar {
+
+    @Override
+    public Map<String, PTransformRunnerFactory> getPTransformRunnerFactories() {
+      return ImmutableMap.of(PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN, new Factory());
+    }
+  }
+  ////////////////////////////////////////////////////////////////////////////////////////////////
+
+  public static <T, W extends BoundedWindow> AssignWindowsRunner<T, W> create(
+      WindowFn<? super T, W> windowFn,
+      Collection<FnDataReceiver<WindowedValue<?>>> fnDataReceivers) {
+    ThrowingConsumer<WindowedValue<T>> consumer =
+        value -> {
+          for (FnDataReceiver receiver : fnDataReceivers) {
+            ((FnDataReceiver<WindowedValue<T>>) receiver).accept(value);
+          }
+        };
+    // Safe contravariant cast
+    WindowFn<T, W> typedWindowFn = (WindowFn<T, W>) windowFn;
+    return new AssignWindowsRunner<>(typedWindowFn, consumer);
+  }
+
+  private final WindowFn<T, W> windowFn;
+  private final ThrowingConsumer<WindowedValue<T>> outputConsumer;
+
+  private AssignWindowsRunner(
+      WindowFn<T, W> windowFn, ThrowingConsumer<WindowedValue<T>> outputConsumer) {
+    this.windowFn = windowFn;
+    this.outputConsumer = outputConsumer;
+  }
+
+  public void assignWindows(WindowedValue<T> compressedInput) throws Exception {
+    for (WindowedValue<T> explodedInput : compressedInput.explodeWindows()) {
+      WindowFn<T, W>.AssignContext ctxt =
```

Comment:
Instead of creating this in each iteration, re-use the same instance and access a reference that is updated within AssignWindowsRunner within the for loop.

Reply:
Added a TODO if this becomes critical.